src/hotspot/share/cds/cppVtables.cpp

```diff
@@ -66,19 +66,20 @@
 
 class CppVtableInfo {
   intptr_t _vtable_size;
-  intptr_t _cloned_vtable[1];
+  intptr_t _cloned_vtable[1]; // Pseudo flexible array member.
 public:
   static int num_slots(int vtable_size) {
     return 1 + vtable_size; // Need to add the space occupied by _vtable_size;
   }
   int vtable_size()           { return int(uintx(_vtable_size)); }
   void set_vtable_size(int n) { _vtable_size = intptr_t(n); }
-  intptr_t* cloned_vtable()   { return &_cloned_vtable[0]; }
-  void zero()                 { memset(_cloned_vtable, 0, sizeof(intptr_t) * vtable_size()); }
+  // Using _cloned_vtable[i] for i > 0 causes undefined behavior. We use address calculation instead.
+  intptr_t* cloned_vtable()   { return (intptr_t*)((char*)this + offset_of(CppVtableInfo, _cloned_vtable)); }
+  void zero()                 { memset(cloned_vtable(), 0, sizeof(intptr_t) * vtable_size()); }
   // Returns the address of the next CppVtableInfo that can be placed immediately after this CppVtableInfo
   static size_t byte_size(int vtable_size) {
     CppVtableInfo i;
-    return pointer_delta(&i._cloned_vtable[vtable_size], &i, sizeof(u1));
+    return pointer_delta(&i.cloned_vtable()[vtable_size], &i, sizeof(u1));
```

Comment:
Rather than making a dummy CppVTableInfo and doing pointer arithmetic, better
would be something like
```
offset_of(CppVtableInfo, _cloned_vtable) + (sizeof(intptr_t) * vtable_size)
```
It might be that some of the subexpressions of that should be broken out into helper
functions that can also be used in `clone_vtable()` and `zero()`.

Also, the really paranoid might `align_up` that to `alignof(CppVtableInfo)`. Currently that's a nop.  Up to you.

Reply:
Right. The pointer_delta variant was not so nice.